src/azul/indexer/log_forwarding_service.py

```diff
@@ -0,0 +1,98 @@
+import csv
+import gzip
+from typing import (
+    Iterator,
+)
+
+from botocore.response import (
+    StreamingBody,
+)
+from furl import (
+    furl,
+)
+
+from azul.deployment import (
+    aws,
+)
+from azul.types import (
+    MutableJSON,
+)
+
+
+class LogForwardingService:
+    """
+    Parse logs from an Application Load Balancer.
+    Adapted from https://github.com/rupertbg/aws-load-balancer-logs-to-cloudwatch
+    """
+
+    @property
+    def s3(self):
+        return aws.s3
+
+    def read_logs(self, bucket: str, key: str) -> Iterator[MutableJSON]:
+        response = self.s3.get_object(Bucket=bucket, Key=key)
+        body: StreamingBody = response['Body']
+        with gzip.open(body, mode='rt') as f:
+            for row in csv.reader(f, delimiter=' ', quotechar='"'):
+                fields = dict(zip(self.fields, row))
+                missing = set(self.fields) - fields.keys()
```

Comment:
This is one of the situations where we need to be a little bit more lenient, or at least handle non-conforming inputs more gracefully. I think the best way to achieve that is to ensure that any exception is logged with the offending S3 key, probably by having a catchall `try: … except:` in `forward_alb_logs`. I don't remember if there is a ticket for the retroactive import or not. Please create one if there isn't. That ticket should also include a metric alarm on any occurrence of failures in `forward_alb_logs` which should produce a very distinctive log signature e.g., by raising a dedicated `LogForwardingException` that the metric filter/alarm can hinge on. The retroactive importer should be able to process a list of keys (in addition to a range of keys).

Reply:
https://github.com/DataBiosphere/azul/issues/4950

Reply:
Note that `pep8` won't allow a bare `except:` with no type exception class.